src/stratigraphy/benchmark/metrics.py

```diff
@@ -0,0 +1,155 @@
+"""Classes for keeping track of metrics such as the F1-score, precision and recall."""
+
+from collections.abc import Callable
+from dataclasses import dataclass
+
+import pandas as pd
+
+
+@dataclass
+class Metrics:
+    """Computes F-score metrics.
+
+    See also https://en.wikipedia.org/wiki/F-score
+
+    Args:
+        tp (int): The true positive count
+        fp (int): The false positive count
+        fn (int): The false negative count
+    """
+
+    tp: int
+    fp: int
+    fn: int
+
+    @property
+    def precision(self) -> float:
+        """Calculate the precision."""
+        if self.tp + self.fp > 0:
+            return self.tp / (self.tp + self.fp)
+        else:
+            return 0
+
+    @property
+    def recall(self) -> float:
+        """Calculate the recall."""
+        if self.tp + self.fn > 0:
+            return self.tp / (self.tp + self.fn)
+        else:
+            return 0
+
+    @property
+    def f1(self) -> float:
+        """Calculate the F1 score."""
+        if self.precision + self.recall > 0:
+            return 2 * self.precision * self.recall / (self.precision + self.recall)
+        else:
+            return 0
+
+
+class DatasetMetrics:
+    """Keeps track of a particular metrics for all documents in a dataset."""
+
+    def __init__(self):
+        self.metrics: dict[str, Metrics] = {}
```

Comment:
Would it make sense for the metric to carry the filename as an attribute? This would make it possible here only to use a file of metrics that already know their filename and avoid creating the dict just to keep track of the filenames. I also had that issue in the refactoring of the metadata metrics and was also not sure about the best solution.

Reply:
I think I'd like to keep the filename separate from the Metrics class for two reasons:
- Currently the Metrics class is also used to compute the metrics for the entire dataset (`DatasetMetrics.overall_metrics()`); in this case there is no specific filename that can be associated with the Metrics object
- In many places where a Metrics object is instantiated, we currently don't have access to the filename. I don't really want to pass additional parameters around, just for this.

That does not mean that we cannot improve the coupling between metrics and filename, but I think it should probably happen at a different level in the code. Maybe we can have a look at it in the context of your refactoring PR?

Reply:
Yes, I tend to share your opinion. Then all good for me!